// @bot-whatsapp/provider/lib/web-whatsapp/index.cjs
/** 
* NO TOCAR ESTE ARCHIVO: Es generado automaticamente, si sabes lo que haces adelante ;)
* de lo contrario mejor ir a la documentacion o al servidor de discord link.codigoencasa.com/DISCORD
*/
'use strict';

var require$$0$1 = require('whatsapp-web.js');
var require$$1$1 = require('@bot-whatsapp/bot');
var require$$2$1 = require('console');
var require$$0 = require('fs');
var require$$1 = require('qr-image');
var require$$2 = require('os');
var require$$3 = require('http');
var require$$4 = require('https');

const { createWriteStream: createWriteStream$1 } = require$$0;
const qr = require$$1;
const { tmpdir } = require$$2;
const http = require$$3;
const https = require$$4;

const cleanNumber$1 = (number, full = false) => {
    number = number.replace('@c.us', '');
    number = !full ? `${number}@c.us` : `${number}`;
    return number
};

const generateImage$1 = (base64) => {
    let qr_svg = qr.image(base64, { type: 'svg', margin: 4 });
    qr_svg.pipe(createWriteStream$1(`${process.cwd()}/qr.svg`));
};

const isValidNumber$1 = (rawNumber) => {
    const regexGroup = /\@g.us\b/gm;
    const exist = rawNumber.match(regexGroup);
    return !exist
};

/**
 * Incompleta
 * Descargar archivo multimedia para enviar
 * @param {*} url
 * @returns
 */
const downloadMedia$1 = (url) => {
    return new Promise((resolve, reject) => {
        const ext = url.split('.').pop();
        const checkProtocol = url.includes('https:');
        const handleHttp = checkProtocol ? https : http;
        const name = `tmp-${Date.now()}.${ext}`;
        const fullPath = `${tmpdir()}/${name}`;
        const file = createWriteStream$1(fullPath);
        handleHttp.get(url, function (response) {
            response.pipe(file);
            file.on('finish', function () {
                file.close();
                resolve(fullPath);
            });
            file.on('error', function () {
                console.log('errro');
                file.close();
                reject(null);
            });
        });
    })
};

var utils = { cleanNumber: cleanNumber$1, generateImage: generateImage$1, isValidNumber: isValidNumber$1, downloadMedia: downloadMedia$1 };

const { Client, LocalAuth, MessageMedia, Buttons } = require$$0$1;
const { ProviderClass } = require$$1$1;
const { Console } = require$$2$1;
const { createWriteStream } = require$$0;
const {
    cleanNumber,
    generateImage,
    isValidNumber,
    downloadMedia,
} = utils;

const logger = new Console({
    stdout: createWriteStream('./log'),
});

/**
 * ⚙️ WebWhatsappProvider: Es una clase tipo adaptor
 * que extiende clases de ProviderClass (la cual es como interfaz para sber que funciones rqueridas)
 * https://github.com/pedroslopez/whatsapp-web.js
 */
class WebWhatsappProvider extends ProviderClass {
    vendor
    constructor() {
        super();
        this.vendor = new Client({
            authStrategy: new LocalAuth(),
        });

        const listEvents = this.busEvents();

        for (const { event, func } of listEvents) {
            this.vendor.on(event, func);
        }
        this.vendor.emit('preinit');
        this.vendor.initialize().catch((e) => {
            logger.log(e);
            this.emit('require_action', {
                instructions: [
                    `(Opcion 1): Debes eliminar la carpeta .wwebjs_auth y reiniciar nuevamente el bot. `,
                    `(Opcion 2): Intenta actualizar el paquete [npm install whatsapp-web.js] `,
                    `(Opcion 3): Ir FORO de discord https://link.codigoencasa.com/DISCORD `,
                ],
            });
        });
    }

    /**
     * Mapeamos los eventos nativos de  whatsapp-web.js a los que la clase Provider espera
     * para tener un standar de eventos
     * @returns
     */
    busEvents = () => [
        {
            event: 'auth_failure',
            func: (payload) => this.emit('error', payload),
        },
        {
            event: 'qr',
            func: (qr) => {
                this.emit('require_action', {
                    instructions: [
                        `Debes escanear el QR Code para iniciar session reivsa qr.svg`,
                        `Recuerda que el QR se actualiza cada minuto `,
                        `Necesitas ayuda: https://link.codigoencasa.com/DISCORD`,
                    ],
                });
                generateImage(qr);
            },
        },
        {
            event: 'ready',
            func: () => this.emit('ready', true),
        },
        {
            event: 'message',
            func: (payload) => {
                if (payload.from === 'status@broadcast') {
                    return
                }

                if (!isValidNumber(payload.from)) {
                    return
                }
                payload.from = cleanNumber(payload.from, true);
                this.emit('message', payload);
            },
        },
    ]

    /**
     * Enviar un archivo multimedia
     * https://docs.wwebjs.dev/MessageMedia.html
     * @private
     * @param {*} number
     * @param {*} mediaInput
     * @returns
     */
    sendMedia = async (number, mediaInput = null) => {
        if (!mediaInput) throw new Error(`NO_SE_ENCONTRO: ${mediaInput}`)
        const fileDownloaded = await downloadMedia(mediaInput);
        const media = MessageMedia.fromFilePath(fileDownloaded);
        return this.vendor.sendMessage(number, media, {
            sendAudioAsVoice: true,
        })
    }

    /**
     * Enviar botones
     * https://docs.wwebjs.dev/Buttons.html
     * @private
     * @param {*} number
     * @param {*} message
     * @param {*} buttons []
     * @returns
     */
    sendButtons = async (number, message, buttons = []) => {
        const buttonMessage = new Buttons(message, buttons, '', '');
        return this.vendor.sendMessage(number, buttonMessage)
    }

    /**
     * Enviar lista
     * https://docs.wwebjs.dev/List.html
     * @private
     * @alpha No funciona en whatsapp bussines
     * @param {*} number
     * @param {*} message
     * @param {*} buttons []
     * @returns
     */
    // sendList = async (number, message, listInput = []) => {
    //     let sections = [
    //         {
    //             title: 'sectionTitle',
    //             rows: [
    //                 { title: 'ListItem1', description: 'desc' },
    //                 { title: 'ListItem2' },
    //             ],
    //         },
    //     ]
    //     let list = new List('List body', 'btnText', sections, 'Title', 'footer')
    //     return this.vendor.sendMessage(number, list)
    // }

    /**
     * Enviar un mensaje solo texto
     * https://docs.wwebjs.dev/Message.html
     * @private
     * @param {*} number
     * @param {*} message
     * @returns
     */
    sendText = async (number, message) => {
        return this.vendor.sendMessage(number, message)
    }

    /**
     *
     * @param {*} userId
     * @param {*} message
     * @param {*} param2
     * @returns
     */
    sendMessage = async (userId, message, { options }) => {
        const number = cleanNumber(userId);
        if (options?.buttons?.length)
            return this.sendButtons(number, message, options.buttons)
        if (options?.media) return this.sendMedia(number, options.media)
        return this.sendText(number, message)
    }
}

var webWhatsapp = WebWhatsappProvider;

module.exports = webWhatsapp;
